ogr/utils.py

```diff
@@ -4,24 +4,76 @@
 import functools
 import logging
 import re
+import time
 from collections.abc import Iterable
 from re import Match
-from typing import Any, Callable, Optional, Union
+from typing import TYPE_CHECKING, Any, Callable, Optional, Union
 
 from urllib3.util import Retry
 
+if TYPE_CHECKING:
+    from urllib3.response import BaseHTTPResponse
+else:
+    # At runtime, BaseHTTPResponse doesn't exist, use HTTPResponse
+    from urllib3.response import HTTPResponse as BaseHTTPResponse
+
 from ogr.abstract import AnyComment, Comment
 
 
 class CustomRetry(Retry):
     """
     Custom Retry class that includes 403 in RETRY_AFTER_STATUS_CODES
     so that Retry-After headers are respected for 403 errors.
+
+    Also handles GitHub rate limit headers (X-RateLimit-Reset) when
+    Retry-After is not present.
     """
 
     # Include 403 in the list of status codes that respect Retry-After header
     RETRY_AFTER_STATUS_CODES = frozenset([413, 429, 503, 403])
 
+    def get_retry_after(self, response: "BaseHTTPResponse") -> Optional[float]:
+        """
+        Override to handle GitHub's X-RateLimit-Reset header.
+
+        GitHub sends X-RateLimit-Reset (Unix timestamp) which we convert
+        to seconds to wait, compatible with Retry-After format.
+
+        This method is called by urllib3's Retry.increment() when
+        respect_retry_after_header=True and the status code is in
+        RETRY_AFTER_STATUS_CODES.
+        """
+        # First, try standard Retry-After header
+        retry_after = super().get_retry_after(response)
+        if retry_after is not None:
+            return retry_after
+
+        # If no Retry-After and this is a rate limit response, check X-RateLimit-Reset
+        if response.status in (403, 429):
+            # GitHub headers are case-insensitive, try both variations
+            rate_limit_reset = response.headers.get(
+                "X-RateLimit-Reset",
+            ) or response.headers.get("x-ratelimit-reset")
```

Comment:
There can be much more variations than just 2, this is not enough to handle them all.

Reply:
I am not sure I follow. You mean I should check for `x-ratelimit-remaining` to be 0? Before handling `x-ratelimit-reset`? Or something else? I was hoping that if we had the reset header than we should use it...

Reply:
See the comment above. The header can be `X-ratelimit-RESET` for all we know.